pennylane/transforms/mitigate.py

```diff
@@ -318,6 +324,46 @@ def richardson_extrapolate(x, y):
     return poly_extrapolate(x, y, len(x) - 1)
 
 
+def exponential_extrapolate(x, y, asymptote=None, eps=1.0e-6):
```

Comment:
In order to make it useful for different machine learning frameworks, you can replace jax.numpy with pennylane qml.math. Then users can use the function with Pytorch, TF, autograd.

Reply:
Cool! Out of curiosity where does `pennylane.math.sign` come from? I can't find it in the [`pennylane/math/__init__.py`](https://github.com/PennyLaneAI/pennylane/blob/48eef6c1c17dc69e5d467a46d5af337c85641fe0/pennylane/math/__init__.py) file.

I haven't figured out to how replace the line
```
y = jnp.array(y)
```
just yet. I tried `qml.math.array` and JAX complained with
```
E       jax.errors.TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on traced array with shape float64[].
E       The error occurred while tracing the function qfunc at /Users/nate/code/pennylane/tests/transforms/test_mitigate.py:459 for jit. This concrete value was not available in Python because it depends on the value of the argument theta.
E       See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerArrayConversionError
E       --------------------
E       For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

/opt/homebrew/Caskroom/miniforge/base/envs/pennylane/lib/python3.11/site-packages/autoray/autoray.py:81: TracerArrayConversionError
```

I **did** get it to work with the following change, but it seems like an "incorrect" way to go about solving this problem.
```diff
-y = jnp.array(y)
+y = qml.math.stack(y)
```

Perhaps you have a recommendation?

Reply:
> Cool! Out of curiosity where does `pennylane.math.sign` come from? I can't find it in the pennylane/math/__init__.py file.

@natestemen we're using [autoray](https://github.com/jcmgray/autoray) to automatically dispatch most of `qml.math.*` to the respective framework :) There are a few we needed to define ourselves either because autoray didn't support them, or there were bugs.

Reply:
Is y not already a Jax array? When you jit the function, you cannot cast abstract type to Jax array. What happens if you do nothing on y?

Reply:
It is not. In `mitigate_with_zne` the results are stored in a python list:
https://github.com/PennyLaneAI/pennylane/blob/3bae4a2c2ea69e7734be8fbeb7641ae7cfe51a52/pennylane/transforms/mitigate.py#L537-L543

Thus, when I try to subtract a float (the asymptote) from the `y` values, I get the following error.
```
>       y_shifted = sign * (y - asymptote)
E       TypeError: unsupported operand type(s) for -: 'list' and 'float'
```

Reply:
Then maybe what would work is to check if it is a list and if it is the case apply array otherwise not. Also you can probably keep the stack, if tests are not failing.

Reply:
I've used `pennylane.math.stack` since it allows us to get rid of the JAX dependency for exponential fitting (https://github.com/PennyLaneAI/pennylane/pull/5972/commits/b60ffc442cc71c1bc6152152edd4638d12dde9e3).

I tested to ensure it works properly on lists and it does, but it's hard to tell what it's doing on more complicated when before `stack` is called we have
```
[Traced<ConcreteArray(2.1213203435596397, dtype=float64)>with<JVPTrace(level=2/0)> with
  primal = Array(2.12132034, dtype=float64)
  tangent = Traced<ShapedArray(float64[])>with<JaxprTrace(level=1/0)> with
    pval = (ShapedArray(float64[]), None)
    recipe = JaxprEqnRecipe(eqn_id=<object object at 0x1749088b0>, in_tracers=(Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>,), out_tracer_refs=[<weakref at 0x17e43c090; to 'JaxprTracer' at 0x17e43c040>], out_avals=[ShapedArray(float64[])], primitive=pjit, params={'jaxpr': { lambda ; a:f64[1]. let
    b:f64[] = reduce_sum[axes=(0,)] a
    c:f64[] = div b 1.0
  in (c,) }, 'in_shardings': (UnspecifiedValue,), 'out_shardings': (UnspecifiedValue,), 'in_layouts': (None,), 'out_layouts': (None,), 'resource_env': None, 'donated_invars': (False,), 'name': '_mean', 'keep_unused': False, 'inline': True}, effects=set(), source_info=SourceInfo(traceback=<jaxlib.xla_extension.Traceback object at 0x1698bce00>, name_stack=NameStack(stack=(Transform(name='jvp'),))), ctx=JaxprEqnContext(compute_type=None,threefry_partitionable=False)), Traced<ConcreteArray(2.120866995454891, dtype=float64)>with<JVPTrace(level=2/0)> with
  primal = Array(2.120867, dtype=float64)
  tangent = Traced<ShapedArray(float64[])>with<JaxprTrace(level=1/0)> with
    pval = (ShapedArray(float64[]), None)
    recipe = JaxprEqnRecipe(eqn_id=<object object at 0x174909000>, in_tracers=(Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>,), out_tracer_refs=[<weakref at 0x17e43c3b0; to 'JaxprTracer' at 0x17e43c360>], out_avals=[ShapedArray(float64[])], primitive=pjit, params={'jaxpr': { lambda ; a:f64[1]. let
    b:f64[] = reduce_sum[axes=(0,)] a
    c:f64[] = div b 1.0
  in (c,) }, 'in_shardings': (UnspecifiedValue,), 'out_shardings': (UnspecifiedValue,), 'in_layouts': (None,), 'out_layouts': (None,), 'resource_env': None, 'donated_invars': (False,), 'name': '_mean', 'keep_unused': False, 'inline': True}, effects=set(), source_info=SourceInfo(traceback=<jaxlib.xla_extension.Traceback object at 0x1698c5970>, name_stack=NameStack(stack=(Transform(name='jvp'),))), ctx=JaxprEqnContext(compute_type=None,threefry_partitionable=False)), Traced<ConcreteArray(2.1206403214025205, dtype=float64)>with<JVPTrace(level=2/0)> with
  primal = Array(2.12064032, dtype=float64)
  tangent = Traced<ShapedArray(float64[])>with<JaxprTrace(level=1/0)> with
    pval = (ShapedArray(float64[]), None)
    recipe = JaxprEqnRecipe(eqn_id=<object object at 0x17e45ed80>, in_tracers=(Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>,), out_tracer_refs=[<weakref at 0x17e43c6d0; to 'JaxprTracer' at 0x17e43c680>], out_avals=[ShapedArray(float64[])], primitive=pjit, params={'jaxpr': { lambda ; a:f64[1]. let
    b:f64[] = reduce_sum[axes=(0,)] a
    c:f64[] = div b 1.0
  in (c,) }, 'in_shardings': (UnspecifiedValue,), 'out_shardings': (UnspecifiedValue,), 'in_layouts': (None,), 'out_layouts': (None,), 'resource_env': None, 'donated_invars': (False,), 'name': '_mean', 'keep_unused': False, 'inline': True}, effects=set(), source_info=SourceInfo(traceback=<jaxlib.xla_extension.Traceback object at 0x16a5c79f0>, name_stack=NameStack(stack=(Transform(name='jvp'),))), ctx=JaxprEqnContext(compute_type=None,threefry_partitionable=False))]
```
and after we have
```
Traced<ConcreteArray([2.12132034 2.120867   2.12064032], dtype=float64)>with<JVPTrace(level=2/0)> with
  primal = Array([2.12132034, 2.120867  , 2.12064032], dtype=float64)
  tangent = Traced<ShapedArray(float64[3])>with<JaxprTrace(level=1/0)> with
    pval = (ShapedArray(float64[3]), None)
    recipe = JaxprEqnRecipe(eqn_id=<object object at 0x17e45eec0>, in_tracers=(Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>, Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>, Traced<ShapedArray(float64[1]):JaxprTrace(level=1/0)>), out_tracer_refs=[<weakref at 0x17e43d210; to 'JaxprTracer' at 0x17e43d1c0>], out_avals=[ShapedArray(float64[3])], primitive=concatenate, params={'dimension': 0}, effects=frozenset(), source_info=SourceInfo(traceback=<jaxlib.xla_extension.Traceback object at 0x1698c8cf0>, name_stack=NameStack(stack=(Transform(name='jvp'),))), ctx=JaxprEqnContext(compute_type=None,threefry_partitionable=False))
```

Reply:
That looks okay to me :+1: